Allow users to choose a different email for notifications

blizzz opened the pull request:
- solves #27465
- part of  #26866

* [x] Implement local email verification
* [x] job to clean up expired verification tokens
* [x] Implement methods to for setting and getting notification emails, as laid out in #27465
* [ ] GUI controls (might be split off)

![Screenshot_20210828_000844](https://user-images.githubusercontent.com/2184312/131193497-e05057f9-a68d-4809-a765-b5ca40aacf7b.png)


![Screenshot_20210820_210826](https://user-images.githubusercontent.com/2184312/130282110-f767668a-2d62-4a94-819b-f46bb9fa8081.png)

### Getting and setting primary mail via provisioning API

`curl -u $userid -X PUT -d 'key=notify_email' -d 'value=myname%40mydomain.com' -H 'OCS-APIRequest: true' https://my.nxt.cld/ocs/v2.php/cloud/users/$userid`

(mind the address must be added as additional email address and also be confirmed)

`curl -u $userid -X GET -H 'OCS-APIRequest: true' https://nc.zara/master/ocs/v2.php/cloud/users/$userid`

blizzz committed:
move verification token logic out of lost password controller

- to make it reusable
- needed for local email verification

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
implement verification for additional mails

- mails added by (sub)admins are automatically verified
- provisioning_api controller as verification endpoint
- IAccountProperty gets a locallyVerified property
- IPropertyCollection gets a method to fetch an IAccountProperty by value
  - an remove equivalent was already present
- AccountManager always initiates mail verification on update if necessary
- add core success template for arbitrary title and message

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix parameter type hint in phpdoc

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
add a job to clean up expired verification tokens

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

LukasReschke commented on apps/provisioning_api/appinfo/routes.php:
```diff
@@ -74,4 +74,8 @@
 	],
+	'routes' => [
+		// Verification
+		['name' => 'Verification#verifyMail', 'url' => '/mailVerification/{key}/{token}/{userId}', 'verb' => 'GET'],
```
Could we go safe here and make this part of a `POST` request?

If you include this link in an email directly, antivirus and anti-spam software will automatically open it. This is currently fine because we check for the user session but I could see someone removing this in the future to make email verifications smoother. (see also https://news.ycombinator.com/item?id=28240279)

blizzz replied:
How would you trigger a POST request from a clickable link send via email?

LukasReschke replied:
GET /verification/foo/bar would be a page that triggers a POST to /verification/foo/bar onload :)

The AV/Antispam software would usually not execute any POST actions on the page.

LukasReschke commented on apps/provisioning_api/lib/Controller/VerificationController.php:
```diff
@@ -0,0 +1,121 @@
+				throw new InvalidArgumentException('Logged in user is not mail address owner');
+			}
+			$email = $this->crypto->decrypt($key);
+			$ref = \substr(hash('sha256', $email), 0, 8);
```
Any specific reason we limit it to the first 8 chars here? :)

blizzz replied:
I consider it sufficient to avoid collisions – this is just part of the configkey stored in the db and used to avoid collisions only. If you manage to craft a collision – mind it applies per user – you manage to overwrite a previously stored token. It could have a security implication, if you manage to create a token for a different user that fits the collision, and yet the payload still needs to be valid and pass the checks.

LukasReschke commented on lib/public/Accounts/IAccountProperty.php:
```diff
@@ -115,4 +115,24 @@ public function setVerificationData(string $verificationData): IAccountProperty;
+	 *
+	 * @since 23.0.0
+	 *
+	 * @param string $verified must be one of the verification constants of IAccountManager
```
I really miss enum support :(

(I guess one could create classes that implement an interface but fair enough)

blizzz replied:
i would say we can have it once PHP 8.1 is the minimum supported version 🚀

LukasReschke reviewed (commented):
Cool stuff and nice work 🎉

blizzz committed:
Confirm mails only per POST

- this is to avoid automatic confirmation by certain softwares that open
  links

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

skjnldsv commented:
/backport to stable22

blizzz committed:
enable the user to set a primary (notification) email address (backend)

- specific getters and setters on IUser and implementation
- new notify_email field in provisioning API

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
use specific email getter where necessary

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
also allow admins to set the primary email

- there will be times when it is necessary to reset this value for sure

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

artonge commented on lib/private/User/User.php:
```diff
@@ -165,24 +169,61 @@ public function setDisplayName($displayName) {
+		}
+
+		$primaryAddress = $this->getPrimaryEMailAddress();
+		if ($primaryAddress === $mailAddress) {
+			// on match no dedicated primary settings is necessary
+			$this->setPrimaryEMailAddress('');
+		}
```
Seems not mandatory to do, maybe this can be removed ?

blizzz replied:
It keeps the DB tidy.

blizzz commented on apps/provisioning_api/lib/Controller/VerificationController.php:
```diff
@@ -0,0 +1,142 @@
+	private $crypto;
+
+	public function __construct(
+		$appName,
```
sure

blizzz commented on apps/provisioning_api/lib/Controller/VerificationController.php:
```diff
@@ -0,0 +1,142 @@
+	 * @NoSubAdminRequired
+	 */
+	public function showVerifyMail(string $token, string $userId, string $key) {
+		if ($this->userSession->getUser()->getUID() !== $userId) {
```
No, auth is necessary, it is not a public page.

blizzz replied:
I'll add a comment

blizzz commented on apps/provisioning_api/lib/Controller/VerificationController.php:
```diff
@@ -0,0 +1,142 @@
+	 */
+	public function verifyMail(string $token, string $userId, string $key) {
+		try {
+			if ($this->userSession->getUser()->getUID() !== $userId) {
```
as above

blizzz commented on lib/private/Accounts/AccountPropertyCollection.php:
```diff
@@ -84,6 +84,15 @@ public function removeProperty(IAccountProperty $property): IAccountPropertyColl
+	public function getPropertyByValue(string $value): ?IAccountProperty {
+		foreach ($this->properties as $i => $property) {
+			if ($property->getValue() === $value) {
+				return $this->properties[$i];
```
It's the same in the end, but I adapt :)

blizzz commented on lib/private/Accounts/AccountPropertyCollection.php:
```diff
@@ -84,6 +84,15 @@ public function removeProperty(IAccountProperty $property): IAccountPropertyColl
 		return $this;
 	}
 
+	public function getPropertyByValue(string $value): ?IAccountProperty {
```
No, it handles properties that may have multiple values, and for now these are just the additional email addresses.

Pytal commented on apps/provisioning_api/lib/Controller/VerificationController.php:
```diff
@@ -0,0 +1,143 @@
+				'title' => $this->l10n->t('Email confirmation'),
+				'message' => $this->l10n->t('To enable the email address %s please click the button below.', [$email]),
+				'action' => $this->l10n->t('Confirm'),
+			], 'guest');
```
We should replace the `'guest'` string with the constant https://github.com/nextcloud/server/blob/9be939300ae0e426d9818756f83f6f09733307fe/lib/public/AppFramework/Http/TemplateResponse.php#L39 and same for the rest of the `TemplateResponse`'s :)

blizzz replied:
done

Pytal reviewed (approved):
Comprehensive 👌
